File: shoggoth/src/arch/x86/src/Idt.cpp
#include "Idt.hpp"
#include "isr.hpp"
#include "irq.hpp"
#include "../../../memory.hpp"

namespace x86 {
    static Idt idt;

    Idt& Idt::getInstance() {
        return idt;
    }

    void Idt::setEntry(int index, void (*isr)(void)) {
        uint32_t addr = (uint32_t) isr;

        memset<uint8_t>(&this->entries[index], 0, sizeof(Entry));
        this->entries[index].offsetLow = addr & 0xFFFF;
        this->entries[index].offsetHigh = addr >> 16;
        this->entries[index].segment = 0x08;
        this->entries[index].type = 0x8E;
    }

    void Idt::init(Arch& arch) {
        volatile Ptr ptr;

        this->setEntry(0, &isr0);
        this->setEntry(1, &isr1);
        this->setEntry(2, &isr2);
        this->setEntry(3, &isr3);
        this->setEntry(4, &isr4);
        this->setEntry(5, &isr5);
        this->setEntry(6, &isr6);
        this->setEntry(7, &isr7);
        this->setEntry(8, &isr8);
        this->setEntry(9, &isr9);
        this->setEntry(10, &isr10);
        this->setEntry(11, &isr11);
        this->setEntry(12, &isr12);
        this->setEntry(13, &isr13);
        this->setEntry(14, &isr14);
        this->setEntry(15, &isr15);
        this->setEntry(16, &isr16);
        this->setEntry(17, &isr7);
        this->setEntry(18, &isr18);
        this->setEntry(19, &isr19);
        this->setEntry(20, &isr20);
        this->setEntry(21, &isr21);
        this->setEntry(22, &isr22);
        this->setEntry(23, &isr23);
        this->setEntry(24, &isr24);
        this->setEntry(25, &isr25);
        this->setEntry(26, &isr26);
        this->setEntry(27, &isr27);
        this->setEntry(28, &isr28);
        this->setEntry(29, &isr29);
        this->setEntry(30, &isr30);
        this->setEntry(31, &isr31);

        //remap the PIC
        arch.getConsole().print("Remapping PIC\n");

        arch.outb(0x20, 0x11);
        arch.outb(0xA0, 0x11);

        //remap
        arch.outb(0x21, 0x20);
        arch.outb(0xA1, 0x28);

        //setup cascading
        arch.outb(0x21, 0x04);
        arch.outb(0xA1, 0x02);

        arch.outb(0x21, 0x01);
        arch.outb(0xA1, 0x01);

        arch.outb(0x21, 0x00);
        arch.outb(0xA1, 0x00);

        //disable the PIC
        arch.getConsole().print("Disabling PIC\n");
        arch.outb(0xA1, 0xFF);
        arch.outb(0x21, 0xFF);

        //setup dummy irq handlers
        for(int count = 0; count < 8; count++) {
            this->setEntry(count + 32, &irqLow);
        }

        for(int count = 0; count < 8; count++) {
            this->setEntry(count + 40, &irqHigh);
        }

        ptr.limit = sizeof(Entry) * 48;
        ptr.addr = (uint32_t) &this->entries[0];

        asm volatile("lidt (%%eax)" : : "a"(&ptr));
    }
}

File: sothoth/kernel/arch/x86/arch/src/idt.rs
use super::isr::*;
use spin::Mutex;

#[repr(C)]
#[repr(packed)]
#[derive(Clone, Copy, Debug)]
struct IdtEntry {
    offset_low_low: u16,
    segment: u16,
    ist: u8,
    attributes: u8,
    offset_low_high: u16,
    offset_high: u32,
    reserved: u32
}

#[repr(C)]
#[repr(packed)]
struct IdtPtr {
    limit: u16,
    addr: u64
}

#[repr(C)]
#[repr(packed)]
struct Idt {
    entries: [IdtEntry; 256]
}

static IDT : Mutex<Idt> = Mutex::new(Idt {
    entries: [IdtEntry {
        offset_low_low: 0,
        segment: 0,
        ist: 0,
        attributes: 0,
        offset_low_high: 0,
        offset_high: 0,
        reserved: 0
    }; 256]
});

pub fn init() {
    IDT.lock().init();
}

impl Idt {
    pub fn init(&mut self) {
        let cs = super::KERNEL_CS;
        let attr : u8 = 0x8E;

        self.set_handler(0, cs, attr, isr0);
        self.set_handler(1, cs, attr, isr1);
        self.set_handler(2, cs, attr, isr2);
        self.set_handler(3, cs, attr, isr3);
        self.set_handler(4, cs, attr, isr4);
        self.set_handler(5, cs, attr, isr5);
        self.set_handler(6, cs, attr, isr6);
        self.set_handler(7, cs, attr, isr7);
        self.set_handler(8, cs, attr, isr8);
        self.set_handler(9, cs, attr, isr9);
        self.set_handler(10, cs, attr, isr10);
        self.set_handler(11, cs, attr, isr11);
        self.set_handler(12, cs, attr, isr12);
        self.set_handler(13, cs, attr, isr13);
        self.set_handler(14, cs, attr, isr14);
        self.set_handler(15, cs, attr, isr15);
        self.set_handler(16, cs, attr, isr16);
        self.set_handler(17, cs, attr, isr17);
        self.set_handler(18, cs, attr, isr18);
        self.set_handler(19, cs, attr, isr19);
        self.set_handler(20, cs, attr, isr20);
        self.set_handler(21, cs, attr, isr21);
        self.set_handler(22, cs, attr, isr22);
        self.set_handler(23, cs, attr, isr23);
        self.set_handler(24, cs, attr, isr24);
        self.set_handler(25, cs, attr, isr25);
        self.set_handler(26, cs, attr, isr26);
        self.set_handler(27, cs, attr, isr27);
        self.set_handler(28, cs, attr, isr28);
        self.set_handler(29, cs, attr, isr29);
        self.set_handler(30, cs, attr, isr30);
        self.set_handler(31, cs, attr, isr31);

        self.load_idt();
    }

    fn set_handler(&mut self, index: usize, segment: u16, attr: u8, handler: unsafe extern fn() -> ()) {
        let addr = handler as u64;

        self.entries[index].offset_low_low = (addr & 0xFFFF) as u16;
        self.entries[index].offset_low_high = ((addr & 0xFFFF0000) >> 16) as u16;
        self.entries[index].offset_high = (addr >> 32) as u32;

        self.entries[index].segment = segment;
        self.entries[index].reserved = 0;
        self.entries[index].attributes = attr;
        self.entries[index].ist = 0;
    }

    fn load_idt(&self) {
        let ptr = IdtPtr {
            limit: 16 * 256,
            addr: (&self.entries[0] as *const IdtEntry) as u64
        };

        unsafe {
            asm!("lidt (%rax)"
            :
            : "{rax}"(&ptr)
            :
            : "volatile");
        }
    }
}
